src/compiler/checker.ts

```diff
@@ -25110,6 +25233,48 @@ namespace ts {
             return literalType;
         }
 
+        /**
+         * Replaces all references to structurally equivalent object and array literal types in the given list
+         * with references to a single one of those types. The process is applied recursively to properties of
+         * object literals and elements of array literals.
+         */
+        function deduplicateObjectOrArrayLiteralTypes(types: Type[]) {
+            if (!some(types, isObjectOrArrayLiteralType)) {
+                return types;
+            }
+            const typeMap = new Map<string, Type>();
+            return sameMap(types, getInternedType);
+
+            function getInternedType(type: Type): Type {
+                if (type.flags & TypeFlags.Union) {
+                    const newTypes = sameMap((<UnionType>type).types, getInternedType);
+                    return newTypes !== (<UnionType>type).types ? getUnionType(newTypes) : type;
+                }
+                const key = getLiteralTypeKey(type);
+                return key ? typeMap.get(key) || (typeMap.set(key, type), type) : type;
+            }
+
+            function getLiteralTypeKey(type: Type): string | undefined {
+                const objectFlags = getObjectFlags(type);
+                if (objectFlags & ObjectFlags.ObjectLiteral) {
+                    const props = getPropertiesOfObjectType(type);
+                    const propTypes = map(props, p => getInternedType(getTypeOfSymbol(p)));

```

Comment:
Does this cause us to be any more eager in evaluating method types than before?

Reply:
No. Method types aren't marked with `ObjectFlags.ArrayLiteral` or `ObjectFlags.ObjectLiteral` so they're not explored and therefore not resolved.